### PR TITLE
week5: update journal

### PR DESCRIPTION
**Pull request by garrettbslone**

### CODE REVIEW

File: journals/week5.md [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+```
+
+And it's clear to see that the job does go thru both the `priority` and `depend` states. My assumption is that the resolution on some combination of the sending of the metrics and aggregating the metrics is not fine enough to mark these transitions as they are very quick. Lastly, why is the new state is getting to around -3.7 million?? Well that is because of two things. First the reason it's so large is because of the way that gauges in brubeck/statsd function. They take and arbitrary *positive* value or a signed value to increment the previous value. The reason that's effecting us is because of the topic string that the `state_cb` was registered under. `job.state.*` will never enter with a current state of `new`, so each time a new job is introduced the plugin catches it on the way out of the `new` state causing the count to be -10,000. Putting those two pieces together, every second when counts are being sent, the `flux.job-states.new` tag is being decremented by 10,000 instead of being set to -10,000. To handle this, I registered the same `state_cb` under the `job.new` topic string which will catch the jobs on the way in and *hopefully* keep the number non-negative.
```
Comment (grondo) [inline journals/week5.md]: Good analysis here!

As jobs are ingested in the job manager, they immediately enter the DEPEND state. If no plugins have emitted `dependency-add` events (i.e. there are no dependencies) then the job will _immediately_ have a `depend` event posted and will proceed to the PRIORITY state. Unless there is a non-default priority plugin loaded, the job will then _immediately_ have a `priority` event posted and will proceed to the SCHED state.

Therefore, in the case of the tests being run here, jobs spend almost no time in the DEPEND and PRIORITY states, but instead proceed immediately to the SCHED state.

File: journals/week5.md [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+
+The last thing I tried was to try adjusting the graphite timing interval by adding a `/opt/graphite/conf/storage-schemas.conf` config file. The following is named flux and it matches any incoming tags that start with flux. For any matching tags it tells graphite that each data point represents 0.1s and we want to keep enough data that it adds up to 1 day. 
+
+```bash
+[flux]
+pattern = ^flux
+retentions = 0.1s:1d
+```
+
+Then I ran another 20,000 jobs under it in two batches of 10,000 to see if it made a difference, and I'm not convinced it is doing what I want. It still seems to be only plotting the data once per minute.
```
Comment (grondo) [inline journals/week5.md:176]: It will be great if we can figure this one out!

File: journals/week5.md [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+
+![job states sched vs inactive](../img/gr_flux_js_is_1.png)
+
+**Cleanup vs Run**
+
+![job states cleanup vs run](../img/gr_flux_js_cr_1.png)
+
+The shorter interval for the watcher had a very minimal performance effect on the jobs themselves, and it did seem to give a little bit more precision wrt to the `sched` and `inactive` states since they are closer to the true starting numbers of 0 and 10,000. It's also interesting to note that this was run on an instance of size 8 so with my 12 core CPU the instance had access to 96 virtual cores, but it never got more than 70 cores at use at once (at least that was reported). It was using on average only about half of the available cores at a time. Then I decided to pick 0.1s as the interval for the timer watcher due mostly to the type being used for the metrics since each time it's called the previous value is just overwritten in the aggregator anyways. Next, I tried adjusting the flush interval within brubeck to 10s and then 1s and running 10,000 jobs in a flux instance for each interval.
+
+**Inactive vs Sched**
```
Comment (grondo) [inline journals/week5.md:154]: Once jobs enter the INACTIVE state, they should never leave it. Any sense on why the count of INACTIVE jobs drops in these graphs? (I'm probably just missing something simple sorry.)

File: journals/week5.md [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+
+### Counting Job States
+
+One interesting place to gather some metrics is from within the job-manager which keeps track of jobs. I set about seeing how to best see how many jobs are in each state at any given time, and also to see how the workload and the cache effected each other. First things first was to get a count of the number of jobs in each state, and the best way to do this is to write a jobtap plugin which hooks into the job-manager module and can be notified upon state transitions. It has a `state_cb` that is invoked upon a `job.state.*` topic which means any job state transition will trigger this callback, and from there its gets the current and previous states for the job and increments the counts both. It also has a timer watcher which is invoked every 1s to send the metrics as gauges using the `fripp_ctx` in the arg. The gauges send a snapshot at the time of whatever the current count of each job state is, so I figured that'd be the best starting point.
```
Comment (grondo) [inline journals/week5.md]: `s/effected each other/affected each other/`

File: journals/week5.md [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+
+**Load/Store Max**
+
+![content cache load/store max](../img/gr_flux_cc_ls_max.png)
+
+**Load/Store Min**
+
+![content cache load/store min](../img/gr_flux_ls_min.png)
+
+A majority of both load and stores were getting completed within 1s, but both did have some large times that pulled up the average for both to >= 2s. The stores tended to be faster, but there were also at least 100x more stores than loads, so there each large load effects the average much more.
```
Comment (grondo) [inline journals/week5.md]: `s/effects the average/affects the average/`

**Review by grondo — APPROVED** [review]
Looks good, thanks again for these detailed entries.

I just made some comments inline, but the only real changes I saw were a couple places where "effect" should be replaced with "affect" 

:+1:

File: journals/week5.md [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+struct fripp_ctx {
+	flux_t *h;
+	struct sockaddr_in si_server;
+	int sock;
+	char *buf;
+	int buf_len;
+};
+```
+
+In turn, that means that the socket is now opened only once per context (when it's being created in `fripp_ctx_create`) and closed once upon destruction of the context, and there is no longer a fixed-size internal buffer for each send (`fripp_format_send`). Instead the buffer in the context is dynamically sized and resized if needed with the idea that most packets being sent will be the same length for a while. One last notable change is that within `fripp_format_send` there is now a second `va_list` (a copy) that's used when the buffer needs to be resized. This was because sometimes, and I'm not sure why it didn't happen every time, upon resizing the buffer the second `vsnprintf` would segfault.
```
Comment (garlick) [inline journals/week5.md:17]: Those va_list bugs can be tricky.  Good sleuthing!

From stdarg(3):
> Each call to `va_arg()` modifies ap  so  that the  next  call returns the next argument. 

and

> If  there  is  no  next argument, or if type is not compatible with the type of the actual next argument (as promoted according to the  default argument promotions), random errors will occur.

So the first call to `vsnprintf()` walks through all the args with `va_arg()`, so in the second call to `vsnprintf()`, there is no next argument and "random errors occur".

C really takes care of you as a programmer...said no one, ever!

File: journals/week5.md [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+```
+
+and then got and sent the final time within the `cache_load_continuation` and `cache_store_continuation` callbacks like:
+
+```c
+fripp_format_send (cache->fripp, "flux.content-cache.store:%lf|ms",
+		   monotime_since (e->fripp_timestamp));
+```
+
+**Load Percentiles**
```
Comment (garlick) [inline journals/week5.md:48]: Might be worth mentioning the Y axis units in the timing graphs.  I'm assuming its milliseconds?

**Review by garlick — APPROVED** [review]
Nice work!

Maybe next week it would be nice to try some graphs with content data and state data on the same graph.  We have some "virtual states" defined in job.h  It might be interesting to see PENDING and RUNNING counts on the same graph as rank 0's count of pending load/store RPCs to the backing store, for example.

**Comment by grondo** [comment]
@garrettbslone - sometimes it is difficult to tell with github: are the comments addressed in this PR yet, and if so is it ready to merge?

**Comment by garrettbslone** [comment]
> @garrettbslone - sometimes it is difficult to tell with github: are the comments addressed in this PR yet, and if so is it ready to merge?

Should be good to merge now.

**Comment by grondo** [comment]
Hm, the `effect` vs `affect` errors still seem to be in the source... are you sure you've force-pushed updates to your branch?

**Comment by grondo** [comment]
Also, one hint for Github PRs -- when you push new commits or force-push changes to your PR branch, reviewers don't necessarily get an email or other notification, so it is good practice to throw in a comment like "force pushed changes to address comments" or something. Thanks!

**Comment by garrettbslone** [comment]
force pushed to fix the `effect` vs `affect`.

**Review by grondo — APPROVED** [review]
LGTM, thanks!!

**Comment by grondo** [comment]
Let me know when you're ready and I'll merge this journal entry. Thanks @garrettbslone!

**Comment by grondo** [comment]
Ah, you already said it was ready so I'll go ahead and  merge.

**Pull request merged**